From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gns <wangbingzhen.riscv@isrc.iscas.ac.cn>
Date: Tue, 11 Aug 2026 20:36:09 +0800
Subject: [PATCH 14/21] riscv(misc): add bytecode listing support

---
 src/jit/bcsave.lua | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/jit/bcsave.lua b/src/jit/bcsave.lua
index 544b506c..4c071335 100644
--- a/src/jit/bcsave.lua
+++ b/src/jit/bcsave.lua
@@ -102,6 +102,7 @@ local map_arch = {
   mips64el =	{ e = "le", b = 64, m = 8, f = 0x80000007, },
   mips64r6 =	{ e = "be", b = 64, m = 8, f = 0xa0000407, },
   mips64r6el =	{ e = "le", b = 64, m = 8, f = 0xa0000407, },
+  riscv64 =    { e = "le", b = 64, m = 243, f = 0x00000004, },
 }
 
 local map_os = {
