From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xiaolin Zhao <zhaoxiaolin@loongson.cn>
Date: Thu, 28 Jul 2022 10:24:50 +0800
Subject: [PATCH 18/91] LoongArch64: Add loongarch64 support when save/list
 bytecode

---
 src/jit/bcsave.lua | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/jit/bcsave.lua b/src/jit/bcsave.lua
index 4c071335..7ce93077 100644
--- a/src/jit/bcsave.lua
+++ b/src/jit/bcsave.lua
@@ -103,6 +103,7 @@ local map_arch = {
   mips64r6 =	{ e = "be", b = 64, m = 8, f = 0xa0000407, },
   mips64r6el =	{ e = "le", b = 64, m = 8, f = 0xa0000407, },
   riscv64 =    { e = "le", b = 64, m = 243, f = 0x00000004, },
+  loongarch64 =	{ e = "le", b = 64, m = 258, f = 0x3, },
 }
 
 local map_os = {
