From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Huang Qiqi <huangqiqi@loongson.cn>
Date: Sat, 12 Oct 2024 10:13:22 +0800
Subject: [PATCH 28/91] LOONGARCH64: Fixed the register allocation bug in asm
 callxs

---
 src/lj_asm_loongarch64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lj_asm_loongarch64.h b/src/lj_asm_loongarch64.h
index abbf5ef6..d18be9b3 100644
--- a/src/lj_asm_loongarch64.h
+++ b/src/lj_asm_loongarch64.h
@@ -324,7 +324,7 @@ static void asm_callx(ASMState *as, IRIns *ir)
   if (irref_isk(func)) {  /* Call to constant address. */
     ci.func = (ASMFunction)(void *)get_kval(as, func);
   } else {  /* Need specific register for indirect calls. */
-    Reg freg = ra_alloc1(as, func, RSET_RANGE(RID_R12, RID_MAX_GPR)-RSET_FIXED);
+    Reg freg = ra_alloc1(as, func, RSET_RANGE(RID_R12, RID_MAX_GPR)&(~RSET_FIXED));
     *--as->mcp = LOONGI_JIRL | LOONGF_D(RID_RA) | LOONGF_J(freg);
     ci.func = (ASMFunction)(void *)0;
   }
