From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Huang Qiqi <huangqiqi@loongson.cn>
Date: Mon, 6 Jan 2025 10:11:59 +0800
Subject: [PATCH 48/91] LOONGARCH64: Fix pcall() error case

---
 src/vm_loongarch64.dasc | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/vm_loongarch64.dasc b/src/vm_loongarch64.dasc
index d4513c04..687910bb 100644
--- a/src/vm_loongarch64.dasc
+++ b/src/vm_loongarch64.dasc
@@ -1328,10 +1328,11 @@ static void build_subroutines(BuildCtx *ctx)
   |  add.d TMP2, BASE, NARGS8:RC
   |  .LDXBU TMP3, DISPATCH, DISPATCH_GL(hookmask)
   |  bltu TMP1, TMP2, ->fff_fallback
-  |  addi.d NARGS8:RC, NARGS8:RC, -8
+  |  addi.d NARGS8:TMP0, NARGS8:RC, -8
   |  or TMP2, BASE, r0
-  |  blt NARGS8:RC, r0, ->fff_fallback
+  |  blt NARGS8:TMP0, r0, ->fff_fallback
   |  addi.d BASE, BASE, 16
+  |  or NARGS8:RC, NARGS8:TMP0, r0
   |  // Remember active hook before pcall.
   |  srli.w TMP3, TMP3, HOOK_ACTIVE_SHIFT
   |  andi TMP3, TMP3, 1
