From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Huang Qiqi <huangqiqi@loongson.cn>
Date: Mon, 13 Jan 2025 15:38:33 +0800
Subject: [PATCH 53/91] LOONGARCH64: Optimize unary test and copy ops

---
 src/vm_loongarch64.dasc | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/src/vm_loongarch64.dasc b/src/vm_loongarch64.dasc
index d1875058..16736e13 100644
--- a/src/vm_loongarch64.dasc
+++ b/src/vm_loongarch64.dasc
@@ -2841,14 +2841,13 @@ static void build_ins(BuildCtx *ctx, BCOp op, int defop)
     |  // RA = dst*8 or unused, RD = src*8, JMP with RD = target
     |  add.d RD, BASE, RD
     |  ld.hu TMP2, OFS_RD(PC)
-    |  ld.d TMP0, 0(RD)
+    |  ld.d CRET1, 0(RD)
     |  addi.d PC, PC, 4
-    |  gettp TMP0, TMP0
+    |  gettp TMP0, CRET1
     |  add.d RA, BASE, RA
     |  sltui TMP0, TMP0, LJ_TISTRUECOND	// TMP0=1 true; TMP0=0 false
     |  decode_BC4b TMP2
     |  addu16i.d TMP3, r0, -0x2		// -BCBIAS_J*4
-    |  ld.d CRET1, 0(RD)
     |  add.w TMP2, TMP2, TMP3		// (jump-0x8000)<<2
     if (op == BC_IST || op == BC_ISTC) {
       |  beqz TMP0, >1
