From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Huang Qiqi <huangqiqi@loongson.cn>
Date: Tue, 14 Jan 2025 19:11:12 +0800
Subject: [PATCH 54/91] LOONGARCH64: Fix the BC_UNM int error

---
 src/vm_loongarch64.dasc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/vm_loongarch64.dasc b/src/vm_loongarch64.dasc
index 16736e13..d50be1e0 100644
--- a/src/vm_loongarch64.dasc
+++ b/src/vm_loongarch64.dasc
@@ -2916,9 +2916,10 @@ static void build_ins(BuildCtx *ctx, BCOp op, int defop)
     |  ld.d TMP0, 0(RB)
     |  addu16i.d TMP1, r0, 0x8000
     |  gettp CARG3, TMP0
+    |  slli.w TMP0, TMP0, 0
     |  bne CARG3, TISNUM, >1
-    |  sub.w TMP0, r0, TMP0
     |  beq TMP0, TMP1, ->vmeta_unm      // Meta handler deals with -2^31.
+    |  sub.w TMP0, r0, TMP0
     |  bstrpick.d TMP0, TMP0, 31, 0
     |  settp TMP0, TISNUM
     |  b >2
