From 430a0a30525937250ff0de09128a570693a1dc82 Mon Sep 17 00:00:00 2001
From: Raymond Wong <infiwang@pm.me>
Date: Tue, 7 Mar 2023 14:31:04 +0800
Subject: [PATCH 067/156] dynasm(riscv): silent compiler warning

---
 dynasm/dasm_riscv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dynasm/dasm_riscv.h b/dynasm/dasm_riscv.h
index 201e98d5..4afe064d 100644
--- a/dynasm/dasm_riscv.h
+++ b/dynasm/dasm_riscv.h
@@ -303,7 +303,7 @@ int dasm_link(Dst_DECL, size_t *szp)
       while (1) {
 	  unsigned int ins = *p++;
 	  unsigned int action = (ins >> 20);
-	  if (ins & 0xf) continue; ins >>= 4;
+	  if (ins & 0xf) continue; else ins >>= 4;
 	  switch (action) {
 	  case DASM_STOP: case DASM_SECTION: goto stop;
 	  case DASM_ESC: p++; break;
-- 
2.42.0

