From 19f80751f25548538c0aac0484f03af0339fd421 Mon Sep 17 00:00:00 2001
From: gns <infiwang@pm.me>
Date: Sun, 25 Jun 2023 14:55:38 +0800
Subject: [PATCH 133/156] riscv(interp): refine BC_MULxx

---
 src/vm_riscv64.dasc | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/vm_riscv64.dasc b/src/vm_riscv64.dasc
index e84dd71d..0f64ea3b 100644
--- a/src/vm_riscv64.dasc
+++ b/src/vm_riscv64.dasc
@@ -3236,12 +3236,10 @@ static void build_ins(BuildCtx *ctx, BCOp op, int defop)
     |  add RA, BASE, RA
     |  bxltz TMP1, ->vmeta_arith
     |.elif "intins" == "mulw"
-    |  mulw CRET1, CARG3, CARG4
     |  mul TMP2, CARG3, CARG4
-    |  sraiw TMP1, CRET1, 31
-    |  srai TMP2, TMP2, 32
+    |  sext.w CRET1, TMP2
     |  add RA, BASE, RA
-    |  bxne TMP1, TMP2, ->vmeta_arith		// 63-32bit not all 0 or 1: overflow.
+    |  bxne CRET1, TMP2, ->vmeta_arith		// 63-32bit not all 0 or 1: overflow.
     |.endif
     |  zext.w CRET1, CRET1
     |  settp CRET1, TISNUM
-- 
2.42.0

