From 230fe159a4df8226ff65d355ee275c80d1d148dc Mon Sep 17 00:00:00 2001
From: gns <infiwang@pm.me>
Date: Wed, 23 Aug 2023 09:08:52 +0800
Subject: [PATCH 156/156] riscv64(asm): fix asm_bswap scratch register list

---
 src/lj_asm_riscv64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lj_asm_riscv64.h b/src/lj_asm_riscv64.h
index 1b4958bf..048254fb 100644
--- a/src/lj_asm_riscv64.h
+++ b/src/lj_asm_riscv64.h
@@ -1380,7 +1380,7 @@ static void asm_bswap(ASMState *as, IRIns *ir)
 {
   Reg dest = ra_dest(as, ir, RSET_GPR);
   Reg left = ra_alloc1(as, ir->op1, RSET_GPR);
-  RegSet allow = rset_exclude(RSET_GPR, dest);
+  RegSet allow = rset_exclude(rset_exclude(RSET_GPR, dest), left);
   if (as->flags & JIT_F_RVZbb) {
     if (!irt_is64(ir->t))
       emit_dsshamt(as, RISCVI_SRAI, dest, dest, 32);
-- 
2.42.0

