diff -up LuaJIT-2.1/etc/luajit.pc.1~ LuaJIT-2.1/etc/luajit.pc
--- LuaJIT-2.1/etc/luajit.pc.1~	2021-07-27 14:50:40.000000000 +0200
+++ LuaJIT-2.1/etc/luajit.pc	2021-08-01 05:42:11.651200382 +0200
@@ -5,7 +5,7 @@ relver=0
 version=${majver}.${minver}.${relver}-beta3
 abiver=5.1
 
-prefix=/usr/local
+prefix=/usr
 multilib=lib
 exec_prefix=${prefix}
 libdir=${exec_prefix}/${multilib}
@@ -20,6 +20,6 @@ Description: Just-in-time compiler for L
 URL: https://luajit.org
 Version: ${version}
 Requires:
-Libs: -L${libdir} -l${libname}
+Libs: -l${libname}
 Libs.private: -Wl,-E -lm -ldl
 Cflags: -I${includedir}
